### ISSUE TITLE
fsdp do not sync gradient for ignored modules

### ISSUE THREAD

**Issue by woolpeeker** [pytorch/pytorch#98281]
### 🐛 Describe the bug

fsdp do not sync gradient for ignored modules

```python
import os
import subprocess
from functools import partial

import torch
import torch.nn as nn
import torch.nn.functional as F

import torch.distributed as dist
from torch.distributed.fsdp.fully_sharded_data_parallel import (
    FullyShardedDataParallel as FSDP,
    ShardingStrategy,
    CPUOffload,
    BackwardPrefetch,
    MixedPrecision,
)

from torch.distributed.fsdp.wrap import (
    size_based_auto_wrap_policy,
    transformer_auto_wrap_policy
)


def get_master_addr():
    _node_list = os.environ['SLURM_JOB_NODELIST']
    _out = subprocess.run(['scontrol', 'show', 'hostnames', _node_list], capture_output=True, text=True)
    return _out.stdout.split('\n')[0]

def slurm_init():
    master_addr = get_master_addr()
    print(f'SLURM_SRUN_COMM_HOST: {master_addr}')
    master_port = int(os.environ.get("MASTER_PORT", "12340"))
    local_rank = int(os.environ["SLURM_LOCALID"])
    world_rank = int(os.environ["SLURM_PROCID"])
    world_size = int(os.environ["SLURM_NTASKS"])

    dist_url = f"tcp://{master_addr}:{master_port}"
    dist.init_process_group(
        backend='nccl',
        init_method=dist_url,
        world_size=world_size,
        rank=world_rank,
    )

    torch.cuda.set_device(local_rank)


def main():
    slurm_init()
    rank = dist.get_rank()
    linear1 = nn.Linear(8, 32, bias=False)
    linear2 = nn.Linear(32, 32, bias=False)
    model = nn.Sequential(
        linear1,
        nn.ReLU(),
        linear2
    ).cuda()

    auto_wrap_policy = partial(size_based_auto_wrap_policy, min_num_params=0)
    mp = MixedPrecision(param_dtype=torch.float32, reduce_dtype=torch.float32, buffer_dtype=torch.float32)
    fsdp = FSDP(model,
         sharding_strategy=ShardingStrategy.FULL_SHARD, 
         auto_wrap_policy=auto_wrap_policy,
         mixed_precision=mp,
         ignored_modules=[linear2],
         sync_module_states=True)
    
    x = torch.randn([1, 8]).cuda()
    y = torch.randn([1, 32]).cuda()

    out = fsdp(x)
    loss = F.l1_loss(out, y)
    loss.backward()
    torch.cuda.synchronize()

    grad = linear2.weight.grad.cpu()
    torch.save(grad, f'grad.{rank}.pth')
    
    dist.barrier()
    
    if rank == 0:
        grad0 = torch.load('grad.0.pth')
        grad1 = torch.load('grad.1.pth')
        abs_err = torch.abs(grad0 - grad1).max()
        print(f"abs_err: {abs_err}")

    dist.barrier()

if __name__ == '__main__':
    main()
```

### Versions

Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 7.5.0
Clang version: Could not collect
CMake version: version 2.8.12.2
Libc version: glibc-2.17

Python version: 3.8.16 (default, Mar  2 2023, 03:21:46)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-3.10.0-693.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 10.2.89
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB
GPU 2: Tesla V100-SXM2-32GB
GPU 3: Tesla V100-SXM2-32GB
GPU 4: Tesla V100-SXM2-32GB
GPU 5: Tesla V100-SXM2-32GB
GPU 6: Tesla V100-SXM2-32GB
GPU 7: Tesla V100-SXM2-32GB

Nvidia driver version: 460.73.01
cuDNN version: /usr/local/cuda-9.0/lib64/libcudnn.so.7.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                48
On-line CPU(s) list:   0-47
Thread(s) per core:    2
Core(s) per socket:    12
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Gold 6146 CPU @ 3.20GHz
Stepping:              4
CPU MHz:               3200.000
BogoMIPS:              6400.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              25344K
NUMA node0 CPU(s):     0-11,24-35
NUMA node1 CPU(s):     12-23,36-47
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 ds_cpl vmx smx est tm2 ssse3 fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm arat pln pts

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==1.12.1
[pip3] torch-ort==1.14.0
[pip3] torchaudio==0.12.1
[pip3] torchvision==0.13.1
[conda] Could not collect

**Comment by pat749** [comment]
The error is caused by the line `_node_list = os.environ['SLURM_JOB_NODELIST']` which is trying to access the environment variable `SLURM_JOB_NODELIST` but it is not defined.
Here is the executable code with expected output: 
```
import subprocess
from functools import partial
import os

import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.nn.parallel import DistributedDataParallel as DDP
import torch.distributed as dist

from torch.distributed.fsdp.fully_sharded_data_parallel import (
    FullyShardedDataParallel as FSDP,
    ShardingStrategy,
    CPUOffload,
    BackwardPrefetch,
    MixedPrecision,
    StateDictType
)

from torch.distributed.fsdp.wrap import (
    size_based_auto_wrap_policy,
    transformer_auto_wrap_policy
)

# import Spur
# from Spur.fsdp import SpurFSDP


def get_master_addr():
    _node_list = os.environ.get('SLURM_JOB_NODELIST')
    if _node_list is None:
        raise ValueError('SLURM_JOB_NODELIST environment variable is not set.')
    _out = subprocess.run(['scontrol', 'show', 'hostnames', _node_list], capture_output=True, text=True)
    return _out.stdout.split('\n')[0]

def slurm_init():
    master_addr = get_master_addr()
    print(f'SLURM_SRUN_COMM_HOST: {master_addr}')
    master_port = int(os.environ.get("MASTER_PORT", "12340"))
    local_rank = int(os.environ["SLURM_LOCALID"])
    world_rank = int(os.environ["SLURM_PROCID"])
    world_size = int(os.environ["SLURM_NTASKS"])

    dist_url = f"tcp://{master_addr}:{master_port}"
    dist.init_process_group(
        backend='nccl',
        init_method=dist_url,
        world_size=world_size,
        rank=world_rank,
    )

    torch.cuda.set_device(local_rank)


def main():
    try:
        slurm_init()
    except ValueError as e:
        print(e)
        return
    
    rank = dist.get_rank()
    linear1 = nn.Linear(8, 32, bias=False)
    linear2 = nn.Linear(32, 32, bias=False)
    model = nn.Sequential(
        linear1,
        nn.ReLU(),
        linear2
    ).cuda()
    if rank == 0:
        torch.save(model.state_dict(), 'model.pth')
    dist.barrier()
    model.load_state_dict(torch.load('model.pth'))

    auto_wrap_policy = partial(size_based_auto_wrap_policy, min_num_params=0)
    mp = MixedPrecision(param_dtype=torch.float32, reduce_dtype=torch.float32, buffer_dtype=torch.float32)
    fsdp = FSDP(model,
         sharding_strategy=ShardingStrategy.FULL_SHARD, 
         auto_wrap_policy=auto_wrap_policy,
         mixed_precision=mp,
         ignored_modules=[linear2])
    
    x = torch.randn([1, 8]).cuda()
    y = torch.randn([1, 32]).cuda()

    out = fsdp(x)
    loss = F.l1_loss(out, y)
    loss.backward()
    torch.cuda.synchronize()

    grad = linear2.weight.grad.cpu()
    print(f"grad.shape: {grad.shape}")
    torch.save(grad, f'grad.{rank}.pth')
    
    dist.barrier()
    
    if rank == 0:
        grad0 = torch.load('grad.0.pth')
        grad1 = torch.load('grad.1.pth')
        abs_err = torch.abs(grad0 - grad1).max()
        print(f"abs_err: {abs_err}")

    dist.barrier()

if __name__ == '__main__':
    main()
```
**OUTPUT:**
```
SLURM_JOB_NODELIST environment variable is not set.
```

Here is the google colab [link](https://colab.research.google.com/drive/1_0BBHbwyLI0VVq3UXPPH0vT0qD7AnV8O?usp=sharing) of this code , please see it . Thank you

**Comment by woolpeeker** [comment]
@pat749 , The environment variable is used to initial the distributed which is required by FSDP. I use the slurm system to start the code with two process, so I write the slurm_init function. You can replace it with yours. Or should I rewrite it to be a more simplier code without it?

**Comment by awgu** [comment]
FSDP is not supposed to reduce the gradient for the ignored module.

**Comment by woolpeeker** [comment]
Without reducing grad for ignored module, it make parameters difference in two GPUs which is wrong for training. Do you have any plan to support it? @awgu

**Comment by awgu** [comment]
Would you be able to explain why you want to ignore the module?

For recsys use cases, we ignore sparse embedding tables, for which we implement our own gradient reduction separately from FSDP. Besides that, we have not seen much motivation for using `ignored_modules` / `ignored_parameters`.

**Comment by awgu** [comment]
> Is there any example codes for the gradient reduction for sparse embedding tables?

I am not sure if there is anything publicly available.

> The ignored module has some code to access the other module's parameters which will trigger error with fsdp.
Changing the code need much efforts involving load checkpoint and optimizer setting. Maybe there are also shared weights I haven't check it.

This makes sense and is an adversarial case for FSDP. Unfortunately, we do not have a great solution for this at the moment, but I will think more from the design side.

**Comment by pengzhenghao-nv** [comment]
> `Would you be able to explain why you want to ignore the module?`

A typical case might be having parameters with different precisions in different submodules of a model. If we don't ignore them, then the parameters can not be flattened.
